nonlinear_solid_mechanics/include/coupling_functions.h

```diff
@@ -0,0 +1,370 @@
+        }
+
+      // pass node coordinates to precice
+      precice.setMeshVertices(precice_mesh_id,
+                              n_interface_nodes,
+                              interface_nodes_positions.data(),
+                              interface_nodes_ids.data());
+
+
+      precice.initialize();
+
+      // write initial writeData to preCICE
+      if (precice.isActionRequired(
+            precice::constants::actionWriteInitialData()))
+        {
+          // store initial write_data for precice in precice_dtp
+          format_deal_to_precice(deal_to_precice);
+
+          precice.writeBlockVectorData(dtp_data_id,
+                                       n_interface_nodes,
+                                       interface_nodes_ids.data(),
+                                       precice_dtp.data());
+
+          precice.markActionFulfilled(
+            precice::constants::actionWriteInitialData());
+
+          precice.initializeData();
+        }
+
+      // read initial readData from preCICE for the first time step
+      if (precice.isReadDataAvailable())
+        {
+          precice.readBlockVectorData(ptd_data_id,
+                                      n_interface_nodes,
+                                      interface_nodes_ids.data(),
+                                      precice_ptd.data());
+
+          format_precice_to_deal(precice_to_deal);
+        }
+    }
+
+
+
+    template <int dim, typename VectorType>
+    void
+    CouplingFunctions<dim, VectorType>::advance_precice(
+      const VectorType &deal_to_precice,
+      VectorType &      precice_to_deal,
+      const double      computedTimestepLength)
+    {
+      if (precice.isWriteDataRequired(computedTimestepLength))
+        {
+          format_deal_to_precice(deal_to_precice);
+          precice.writeBlockVectorData(dtp_data_id,
+                                       n_interface_nodes,
+                                       interface_nodes_ids.data(),
+                                       precice_dtp.data());
+        }
+
+      precice.advance(computedTimestepLength);
+
+      if (precice.isReadDataAvailable())
+        {
+          precice.readBlockVectorData(ptd_data_id,
+                                      n_interface_nodes,
+                                      interface_nodes_ids.data(),
+                                      precice_ptd.data());
+
+          format_precice_to_deal(precice_to_deal);
+        }
+    }
+
+
+    // TODO: Check this again, especially operator[] for BlockVectors
+    template <int dim, typename VectorType>
+    void
+    CouplingFunctions<dim, VectorType>::format_deal_to_precice(
+      const VectorType &deal_to_precice)
+    {
+      int data_iterator = 0;
+      for (auto element = coupling_dofs.begin(); element != coupling_dofs.end();
+           ++element)
+        {
+          precice_dtp[data_iterator] = deal_to_precice[*element];
+
+          ++data_iterator;
+        }
+    }
```

Comment:
This appears unnecessary complicated. Are both vectors needed at all?

Reply:
Yes, one is an actual deal.II vector, sorted according to the dof numbering holding all dofs of the problem of deal.II and one is a sorted `std::vector`, which only holds the relevant coupling data for precice.